challenge 2 (#16)
fermyon/summer-of-spin: challenge-2/README.md

# Challenge 2
Inspired from [Challenge 3 of Advent of Spin 2023](https://github.com/fermyon/advent-of-spin/blob/main/2023/Challenge-3/README.md
)

> 

Author: Yep! I will remove the inspiration from the challenge 3 part as well.

### Spec
It's summer time! It's all about vacations and enjoying the beautiful weather outdoors. Finally, a group of friends found time to take a trip that they have been putting off for months, Alas! They don't want to deal with planning every part of the trip. Your task is to utilize the LLM powers to help them build a perfect itinerary to enjoy the season. This must be stored in a key-value store for them to refer later as they go on this trip! 

The output generated by the model is not assessed but the http response is, as discussed below.

The users will be able to give some input such as destination, duration, type of trip and number of people. Take the help of LLMs to plan the perfect getaway! The users will also provide a unique `tag` to retrieve the generated result later.

The users will send a POST request to `/plan-my-trip` in JSON format.
```json
{
    "tag": "week-in-banff"
    "destination": "Banff",
    "duration": "1 week",
    "num_people": "four",
    "activities": ["hiking", "camping", "long drive", "food"]
    
}
```

> You will have to create a prompt out the JSON data given along with the request to feed it to the LLM model to generate the itinerary

Reviewer: ```suggestion
You will have to create a prompt from the JSON data in the request and feed it to the LLM model to create an itinerary.
```

Example prompt template might be

```
PROMPT = " Create a summer vacation detailed itinerary trip to go to <destination> for a <duration>. <num_people> people will be going on this trip planning to do <activities>"
```
Do explore different ways to generate the best outcome from the model :D

The expected return type must be a `201` response as follows:

```json
{
    "itinerary" : <GENERATED ITINERARY>

> }

Reviewer: You probably want the backend to generate the tag and not the user providing it to avoid one user overwritting another.

```
The header type must contain `Content-Type: application/json`

> For reference, the users will have to send a `GET` request to route `\<tag>` given along with the previous `POST` request to see the generated itinerary again. Store the itinerary in a KV Store for easier retrieval in `tag:<GENERATED ITINERARY>` format. When requested the response should return with a `200` response code in the following way: 

Reviewer 2: I am having a hard time understanding this section. You seem to be mixing together the user experience description with the backend implementation description. Can you work on rewording it? Maybe move the section about storing the result up to where you talk about the return response, telling them to persist the result before returning 201. Here is a start:
```suggestion
To reference a previously generated itinerary, a user can send a `GET` request to the same endpoint with the  `/<tag>` route. The itinerary is then retrieved from the key-value store where it was cached using its tag. The response should have a `200` response code and be formatted as follows: 
```

```json
{
    "itinerary": <GENERATED ITINERARY>
}
```

> Don't forget to put your "spinscreen" on (no pun intended) and all the best! :D

Reviewer 2: Pun is intended :)
```suggestion
Don't forget to put your "spinscreen" on (pun intended)! Happy vacation! :D
```
